RedHatInsights/rhsm-subscriptions
# ENT-3840: Refactor metering job to be more generic

Primarily, this adds to the task messages two fields:

- `productProfileId` (which profile is being metered)
- `metric` (which metric is being collected, identified by unit of measure for now)

I added to the JMX methods `productProfileId` as a parameter, and modified the PrometheusMetricsTaskManager methods to loop through metrics for a given product and queue up tasks for each metric

Note that I did placeholder implementations of some methods in `PrometheusMetricsProperties`, these should be refactored and/or removed in ENT-3835.

Note the following renames:

- OpenShiftJmxBean -> MeteringJmxBean
- OpenShiftMeteringJob -> MeteringJob
- OpenShiftJmxProfile -> MeteringJmxProfile
- OpenShiftJobProfile -> MeteringJobProfile
- OpenShiftTasksConfiguration -> MeteringTasksConfiguration
- OpenShiftMetricsTask -> MetricsTask

I also:

- Renamed the `openshift-metering-job` profile to `metering-job`.
- Renamed the `openshift-metering-jmx` profile to `metering-jmx`.
- Switched the topic name from `platform.rhsm-subscriptions.openshift-metering-tasks` to `platform.rhsm-subscriptions.metering-tasks`

Testing
-------

1. Run the sidecar for token-refresher against staging environment (ask internally for details if needed).
2. Run the server with `PROM_URL="http://localhost:8082/api/v1" USER_USE_STUB=true ./gradlew bootRun`.
3. Queue up an account to be sync metrics for:

```
curl 'http://localhost:8080/actuator/jolokia/' \
  -H 'Content-Type: application/json' \
  -d '{"type":"exec","mbean":"org.candlepin.subscriptions.metering.jmx:name=meteringJmxBean,type=MeteringJmxBean","operation":"performCustomMetering(java.lang.String,java.lang.String,java.lang.Integer)","arguments":["OpenShift","2021-05-19T09:00:00Z",60]}'
```

4. Verify that the promtheus call is still attempted.
